dss/util/iterators.py

```diff
@@ -0,0 +1,177 @@
+    └───┴───┼───┤           ├───┼───┬───┐  Values not   ├───┼───┼───┼───┤    denote
+            │ 3 │           │ 2 │ 2 │ 2 │   equal to    │ 2 │ 2 │ 2 │ 2 │    None)
+            └───┘           └───┴───┼───┤    minimum    ├───┼───┴───┼───┤
+                                    │ 3 │   are to be   │ 3 │       │ 3 │
+                                    └───┘    ignored    └───┘       └───┘
+
+    While the center representation is the most intuitive one, zipalign returns the one on the right as it allows for
+    the elimination of intrinsic state and more efficient execution. It is also richer in that it distinguishes
+    between cells missing from an iterator (value != min) and cells past the end of an iterator (value is None).
+
+    >>> i = zipalign(columns)
+    >>> next(i)
+    Row(min=0, values=(1, 0, 0))
+    >>> next(i)
+    Row(min=1, values=(1, 2, 2))
+    >>> next(i)
+    Row(min=2, values=(2, 2, 2))
+    >>> next(i)
+    Row(min=3, values=(None, None, 3))
+    >>> next(i)
+    Traceback (most recent call last):
+    ...
+    StopIteration
+
+    Converting the actual result back to the more intuitive representation is straight-forward …
+
+    >>> [tuple(val if val == row.min else None for val in row.values) for row in zipalign(columns)]
+    [(None, 0, 0), (1, None, None), (2, 2, 2), (None, None, 3)]
+
+    … and row objects offer a convenience method for that:
+
+    >>> [row.norm() for row in zipalign(columns)]
+    [(None, 0, 0), (1, None, None), (2, 2, 2), (None, None, 3)]
+
+    A few more properties: For one, exhaustion is permanent:
+
+    >>> next(i)
+    Traceback (most recent call last):
+    ...
+    StopIteration
+
+    If any input iterators violate the ordering constraint, the behavior is well-defined: an exception is raised as
+    soon as the out-of-order element is detected.
+
+    >>> list(zipalign([(1, 0)]))
+    Traceback (most recent call last):
+    ...
+    ValueError: Input iterator yielded value out of order
+
+    Given a Row object, a zipalign iterator can resume where another zipalign iterator left off:
+
+    >>> columns = list(map(iter, columns))
+    >>> row = next(zipalign(columns)); row
+    Row(min=0, values=(1, 0, 0))
+    >>> next(zipalign(columns, row=row))
+    Row(min=1, values=(1, 2, 2))
+
+    Resumption also works with a plain iterable instead of a Row instance …
+
+    >>> next(zipalign(columns, row=[1, 2, 2]))
+    Row(min=2, values=(2, 2, 2))
+
+    … provided that its length is equal to the number of columns.
+
+    >>> next(zipalign(columns, row=[1, 2, 2, 3]))
+    Traceback (most recent call last):
+    ValueError: Row length (4) does not match # of columns (3)
+
+    We can resume at any stage …
+
+    >>> next(zipalign(columns, row=[2, 2, 2]))
+    Row(min=3, values=(None, None, 3))
+
+    … even at the last row.
+
+    >>> next(zipalign(columns, row=[None, None, 3]))
+    Traceback (most recent call last):
+    ...
+    StopIteration
+    """
+
+    class Row(NamedTuple):
```

Comment:
Good catch.

Reply:
though I suppose `NamesTuple`s are pretty optimized already. If it's simple change try using `__slots__`, otherwise we can optimize it if it become a problem.

Reply:
Actually, NamedTuple prohibits defining __slots__:

AttributeError: Cannot overwrite NamedTuple attribute __slots__

Digging deeper, `collections.namedtuple` already sets __slots__ to an empty tuple. `typing.NamedTuple` uses `collections.namedtuple` under the hood so the same applies to classes derived from `typing.NamedTuple`. You only need to set `__slots__` if you derive from a derivation of `typing.NamedTuple`:

```
>>> from typing import NamedTuple
>>> class Foo(NamedTuple):
...     x: str
...     y: int
... 
>>> Foo('s',2).__dict__
Traceback (most recent call last):
  File "<input>", line 1, in <module>
AttributeError: 'Foo' object has no attribute '__dict__'
>>> class Bar(Foo):
...     pass
... 
>>> Bar('s',2).__dict__
{}
>>> class Bar2(Foo):
...     __slots__ = ()
...     
>>> Bar2('s',2).__dict__
Traceback (most recent call last):
  File "<input>", line 1, in <module>
AttributeError: 'Bar2' object has no attribute '__dict__'
```